### PR TITLE
Fix backoffice JS code, fix doubled and tripled pageviews, fix echo in hook, fix default BO tracking - for versions up to 1.7.6

### PR DESCRIPTION
**Pull request by Hlavtox**
# This PR fixes following bugs and confusions

### JS code that was killed before page output
- JS code that is not included in a backoffice header because of https://github.com/PrestaShop/ps_googleanalytics/pull/72/. This PR changes the logic and disables only the pageview itself in the TPL.
- This logic needs to work correctly because other functionality depends on it.
-- Default actionProductCancel functionality of the module
-- Order refund functionality I am adding here: https://github.com/PrestaShop/ps_googleanalytics/pull/66

### Backoffice views enabled by default
- Until 4.0.0, this module did not track backoffice views.
- There was a bug in 4.0.0 that enabled backoffice pageviews to be sent.
- @matks decided to keep the tracking there and disable it when wanted.
- Based on previous behavior, the default state should be that the module should not track backoffice views. 
- If users upgrade to version 4, they would not be happy that it's behaving differently without their knowledge and they have to go to module settings to disable it.

### Doubled and tripled pageviews
- Removes a code that sends two or three extra pageviews in BO. This code is not present in v3. Pageview is sent in the header, no need to have it in the other TPL. This code was not present in v3 of the module.

### Code output to order admin page was not working
- Fixes an echo in one hook. 
- Echo does not work on migrated pages and needs to be changed to return.

### How to test
- Install Google Tag Assistant
- Try tracking in FO - _properly included, 1 pageview_
- Try tracking in BO with BO tracking enabled - _properly included, 1 pageview will be sent_
- Try tracking in BO with BO tracking disabled - _properly included, no pageview is sent (reports no HTTP response, that's normal)_
- Try triggering hookActionProductCancel, refund code will be output on BO order view page, if needed.

### CODE REVIEW

File: ps_googleanalytics.php
Comment (kpodemski) [inline ps_googleanalytics.php:142]: @Hlavtox do we really need this empty var here?

Comment (Hlavtox) [reply ps_googleanalytics.php:142]: @kpodemski Maybe not?

If the function somehow fails, is it ok if we return false or whatever?

**Comment by kpodemski** [comment]
Didn't test it but looks good for me 👍 

I'm not sure why we should allow pageViews in back-office in the first place?

File: ps_googleanalytics.php
Comment (kpodemski) [reply ps_googleanalytics.php:142]: @Hlavtox yes, unless there's a check `if (Hook::exec)` but that's not the case in PrestaShop from what I know :)

Comment (kpodemski) [reply ps_googleanalytics.php:142]: @Hlavtox you can check this by having `return false;` here and creating other module which is also hooked into `displayAdminOrder`, just to make sure ;)

Comment (Hlavtox) [reply ps_googleanalytics.php]: Removed ;-)

**Comment by Progi1984** [comment]
Hi @Hlavtox, before going inside this PR, we need to fix this issue #74 and create specs for this module. If you are interested, you can open a PR on the repository [prestashop-specs](https://github.com/PrestaShop/prestashop-specs/tree/master/modules).

**Comment by Hlavtox** [comment]
@Progi1984 I do not change any functionality. I restore the compatibility for other logic.

**Comment by Hlavtox** [comment]
@Progi1984 Can you check the PR again? Why we need to wait a century until somebody writes a module spec when I do not change the the module behavior?

I only repair what somebody f***ed up. 🤷‍♂️

File: classes/Form/ConfigurationForm.php
Comment (eternoendless) [inline classes/Form/ConfigurationForm.php:179]: Changing this settings requires an [upgrade script](https://devdocs.prestashop.com/1.7/modules/creation/enabling-auto-update/) so that settings from previous versions (`GA_TRACK_BACKOFFICE_DISABLED`) are cleaned up an carried over to `GA_TRACK_BACKOFFICE_ENABLED` in the new version.

**Review by eternoendless — CHANGES_REQUESTED** [review]
Could you please explain why the GA script must be initialized even when you choose not to track in BO?

**Comment by ttoine** [comment]
BTW, @Hlavtox, many people have contacted me about a couple of your recent posts, including one in this discussion.

So, could you please re-read the [Code of Conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)? And then, try to adapt your tone, so discussions could be respectful and constructive, without aggressiveness? Because, if you continue like that, I guess that some people will fear to collaborate with you, even if your contribution to the project is interesting.

File: classes/Form/ConfigurationForm.php
Comment (kpodemski) [reply classes/Form/ConfigurationForm.php:179]: Previous version was never released, is it needed in that case?

Comment (Hlavtox) [reply classes/Form/ConfigurationForm.php:179]: @eternoendless The setting only made it to the dev version, no need to do an upgrade script I guess?

**Comment by Hlavtox** [comment]
@eternoendless 

`Could you please explain why the GA script must be initialized even when you choose not to track in BO?`
Because there is a refund functionality in the module that outputs JS code in the middle of the page. Imagine something like `ga(action = 'refund', product = '123');`
In the way @matks has done it, this will stop working, because the JS code is killed before executing the header hook. BO tracking option should only affect the pageviews, not the entire script being blocked.

**Comment by Hlavtox** [comment]
@ttoine Ok, let's go through it again.

- I am adding a order refund functionality to Google Analytics.
- This takes a century to review.
- In the meantime, you completely refactor the module and change everything.
- OK, I redo everything.
- Somebody messes up and the module suddenly tracks backoffice views.
- Somebody messes up even more and completely kills the JS code my PR relies on. When I tell him several times that it cannot be done that way, my comments and messages are ignored.
- OK, let's fix it.
- I make a PR with all the fixes and @Progi1984 that my PR needs to wait until you write some kind of module specifications? With the speed everything takes at Prestashop it will take a year. And what kind of specifications you need to write for Google Analytics module? Just do it as Google says.

**Comment by ttoine** [comment]
I understand, I guess we all understand the frustration. Unfortunately, like in many other open source projects, this is happening here too. (not too often, thankfully). It happened before to other people in the past. And, it will happen again, for sure.

But being agressive and using insults will not help at all. It will just make people less motivated to collaborate with you.

File: classes/Form/ConfigurationForm.php
Comment (eternoendless) [reply classes/Form/ConfigurationForm.php:179]: Oh in that case it's not needed. I thought it was already released.

**Review by eternoendless — APPROVED** [review]
Okay thanks for the explanation. Changes look good according to what you said. I'm not familiar enough with this module to understand the implications, WDYT @matks ?

**Comment by Hlavtox** [comment]
@eternoendless Thank you! :-)

**Comment by eternoendless** [comment]
BTW the reason @Progi1984 asks for specifications is that since we don't know how the module is expected to behave, it's very hard to validate changes. This is unfortunately the case in many places throughout the project because previous maintainers didn't leave a "paper trail", so to speak, to allow other people to pick up the job afterwards. So we're left with reverse engineering and guesswork, which obviously leads to many, may problems.

This is changing, but as you know already, it also takes a lot of time.

**Comment by Hlavtox** [comment]
@eternoendless I perfectly understand it, but I would not hold off a fix that could upset a lot of people because of it. :-(

Imagine how many users will just click upgrade in their Prestashop BO and now they will have some mess in their statistics. Most visited page will now suddenly be BO orders... 😄

And regarding the specs, as I said, I think it's pretty easy in case of this module, isn't it. There is not much to change or alter, if the implementation is done per Google's documentation. [https://developers.google.com/analytics/devguides/collection/analyticsjs/enhanced-ecommerce#measuring-transactions](https://developers.google.com/analytics/devguides/collection/analyticsjs/enhanced-ecommerce#measuring-transactions)

Maybe some users want to do some non-standard things (send amount with taxes, where it's supposed to be without etc.), but that's their responsibility.

**Comment by eternoendless** [comment]
> And regarding the specs, as I said, I think it's pretty easy in case of this module, isn't it

Great! I suppose that means you can write them! (just kidding, but if you did that it would be awesome 😄)

> And when I post some advice and somebody just completely ignores it, it just makes me think that the person thinks I'm stupid and incompetent

Oh no, rest assured, that's absolutely not the case. It's that we're DDoS by dozens of things every day and we can't invest as much time as we'd like in everything. That said, the fact that you're active in the project and that you point out nontrivial stuff has made you known to the team and, if anything, be _less_ "ignored" than others.

**Comment by matks** [comment]
Hi @Hlavtox thank you for the detailed explanations about this PR

> # This PR fixes following bugs and confusions
>
> ### JS code that was killed before page output
> * JS code that is not included in a backoffice header because of #72. This PR changes the logic and disables only the pageview itself in the TPL.
> * This logic needs to work correctly because other functionality depends on it.
>   -- Default actionProductCancel functionality of the module
>   -- Order refund functionality I am adding here: #66

Approved, it's clean ✅ . If we get QA validation, it's a yes.

> ### Backoffice views enabled by default
> * Until 4.0.0, this module did not track backoffice views.
> * There was a bug in 4.0.0 that enabled backoffice pageviews to be sent.
> * @matks decided to keep the tracking there and disable it when wanted.
> * Based on previous behavior, the default state should be that the module should not track backoffice views.
> * If users upgrade to version 4, they would not be happy that it's behaving differently without their knowledge and they have to go to module settings to disable it.

This is discutable. v4.0.0 is a new version that does bring in new features. We can see, when looking at the code, that _this new 'tracking in BO' feature was wanted_.

The reason why I enabled the tracking in BO per default was to ensure that order cancel or refund events would be sent this way. If we follow your changes, indeed there is no need to enable tracking by default.

> ### Doubled and tripled pageviews
> * Removes a code that sends two or three extra pageviews in BO. This code is not present in v3. Pageview is sent in the header, no need to have it in the other TPL. This code was not present in v3 of the module.

I think you refer to the code statement in `views/templates/hook/ga_tag.tpl` right ?

Without better exploration of the code and [without documentation / specifications](https://github.com/PrestaShop/ps_googleanalytics/issues/74) I don't think we can simply remove this code. We cannot simply remove something without being 100% sure it is not useful. Else we might end up in [nasty situations](https://pbs.twimg.com/media/EirQkKhXYAM8A57?format=jpg&name=small). Moreover this is open source, people who use it are usually against removing things from the software.

We can explore this further to validate that this part of the code must go.

This is not something I write to annoy you. _We need to be extra careful when removing something_ because we don't know the inner workings of this module. As explained by @eternoendless because previous maintainers didn't leave a "paper trail", we're left with reverse engineering and guesswork, and this is not reliable.

So **if we can assert with a trust percentage high enough that this code should be removed** I'll approve the PR. I will not approve this part of the PR without this assertion. Once again, _extra caution is need here_. It's no fun, it's long, bothersome but this is the only safe way for a better release.

> ### Code output to order admin page was not working
> * Fixes an echo in one hook.
> * Echo does not work on migrated pages and needs to be changed to return.

Approved, it's clean ✅ . If we get QA validation, it's a yes.

## Summary of 'blocking items'

1. Need an exploration of the code and usecases to make sure code `views/templates/hook/ga_tag.tpl` can be removed
2. Need to make sure we're all OK with disabling tracking in Back-Office by default

**Comment by matks** [comment]
@Hlavtox After checking with multiple stakeholders, I encountered no issues about putting the BO tracking on opt-in (= disabled by default ; enabled when merchant chooses to)

**Comment by kpodemski** [comment]
I wonder what's the point of sending pageview from back-office, isn't it mistake in the first place?

**Comment by matks** [comment]
> I wonder what's the point of sending pageview from back-office, isn't it mistake in the first place?

I wonder too, but if you look at this code:
https://github.com/PrestaShop/ps_googleanalytics/blob/master/views/templates/hook/ga_tag.tpl#L32
```
{if ($jsState != 1 && $isBackoffice === true)}
    {literal}
    <script type="text/javascript">
        ga('send', 'pageview');
    </script>
    {/literal}
{/if}
```

Unless it's very badly written, the intention is clearly "when we are in BO and this weird `jsState` is set to 1, send a pageview".

**Comment by matks** [comment]
Ok, after digging into the git history I think I found the root cause

In commit https://github.com/PrestaShop/ps_googleanalytics/commit/3ec09fea2a1afa02c55402e5426eef4796c78d59#diff-74214ee2c66b22a15e9ec506a2e0b1af
the script is
```
{if ($jsState != 1 && $isBackoffice == 0)}
    {literal}
    <script type="text/javascript">
        ga('send', 'pageview');
    </script>
    {/literal}
{/if}
```
So we can see it's not intended for backoffice usage

But later commit https://github.com/PrestaShop/ps_googleanalytics/commit/07563209a788a151730fc2ff5f3a678d7af47bb9#diff-74214ee2c66b22a15e9ec506a2e0b1af
change
```
{if ($jsState != 1 && $isBackoffice == 0)}
```
into
```
{if ($jsState != 1 && $isBackoffice === true)}
```

**Comment by kpodemski** [comment]
Oh, nice catch, so it was a mistake.

I think that even tho this module seems to be simple, having few points of how certain things in module should work is a must have... 

I'm pretty sure that this would be eventually highly appreciated in next refactoring of the module.

**Comment by matks** [comment]
@Hlavtox Regarding this

> ### Doubled and tripled pageviews
> * Removes a code that sends two or three extra pageviews in BO. This code is not present in v3. Pageview is sent in the header, no need to have it in the other TPL. This code was not present in v3 of the module.

After digging through git history, we can now establish that this code did exist in v3 of the module. It was there https://github.com/PrestaShop/ps_googleanalytics/blob/3.1.3/ps_googleanalytics.php#L713 . It was code initially being built inside the main module file `ps_googleanalytics.php` and then inserted into the page.

However pageview is, in v4, being added through https://github.com/PrestaShop/ps_googleanalytics/blob/master/views/templates/hook/ps_googleanalytics.tpl which deprecates the usage of this file. This, combined with the condition inversion, results in the current situation.

I believe that these new elements allow us to assert with a trust percentage high enough that this code should be removed. In other words, I approve the Pull Request intent as these changes are both valid and desireable for the module behavior.

There is however something that might be wrong:

**Review by matks — APPROVED** [review]

**Comment by Hlavtox** [comment]
@matks @kpodemski @eternoendless 

Guys, thank you for going through it and approving it!

_____________

Now, for you to have a calm sleep... :-D I think that on 99 %, the BO tracking was not intended. If you look on the following code. https://github.com/PrestaShop/ps_googleanalytics/blob/master/views/templates/hook/ps_googleanalytics.tpl

```
{if $backOffice}
  ga('set', 'nonInteraction', true);
{else}
  ga('send', 'pageview');
{/if}
```
The pageview was sent because $backoffice variable is not working, not because somebody wanted to be sent.

But, good thing in the end we fixed it and now it has yet another feature for people who would like to use it. 🚀

**Review by PierreRambaud — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Comment by Hlavtox** [comment]
bump pls, people are getting nonsense in their GA accounts @sarahdib

**Comment by SD1982** [comment]
Hi @Hlavtox With this PR I have an error in the order view when i cancel a product
image

**Comment by matks** [comment]
Hi @SD1982 I checked this error, I also have it with v3.3.0 and v4.0.0 this is not a regression. This is a compatibility issue between the module and PS 1.7.7.0. I think if you check on PS 1.7.6.9 it works.

**Comment by SD1982** [comment]
Hi @Hlavtox as said [here](https://github.com/PrestaShop/ps_googleanalytics/pull/77#issuecomment-731644392) by matks and after more tests the PR seems ok but only for 1.7.6.9 for now.

**Comment by Hlavtox** [comment]
@SD1982 Ok so you can merge it? The bug with cancelling products is not my problem I guess? 😛

**Comment by matks** [comment]
Yes we can merge it 👍  because PS 1.7.6.9 is the current latest stable version

**Comment by matks** [comment]
I added in the title the fact that it does support until PS 1.7.6

Compatibility issue for PS 1.7.7 is reported here https://github.com/PrestaShop/ps_googleanalytics/issues/84

**Comment by Hlavtox** [comment]
Thanks @matks @SD1982 🚀

**Pull request merged**
